src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -96,6 +103,38 @@ public static ImmutableArray<DocumentId> GetDocumentIds(this Solution solution,
                 : documents.FindDocumentInProjectContext(documentIdentifier, (sln, id) => sln.GetRequiredDocument(id));
         }
 
+        /// <summary>
+        /// Finds the document for a TextDocumentIdentifier, potentially returning a source-generated file.
+        /// </summary>
+        public static async ValueTask<Document?> GetDocumentAsync(this Solution solution, TextDocumentIdentifier documentIdentifier, CancellationToken cancellationToken)
+        {
+            if (documentIdentifier.Uri.Scheme == SourceGeneratedDocumentUris.Scheme)
+            {
+                // In the case of a URI scheme for source generated files, we generate a different URI for each project, thus this URI cannot be linked into multiple projects;
+                // this means we can safely call .SingleOrDefault() and not worry about calling FindDocumentInProjectContext.
+                var documentId = solution.GetDocumentIds(documentIdentifier.Uri).SingleOrDefault();
+                return await solution.GetDocumentAsync(documentId, includeSourceGenerated: true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return solution.GetDocument(documentIdentifier);
+            }
+        }
+
+        public static async ValueTask<TextDocument?> GetTextDocumentAsync(this Solution solution, TextDocumentIdentifier documentIdentifier, CancellationToken cancellationToken)
```

Comment:
I'm getting a bit confused between all these overloads we have.  Should `GetDocumentAsync` be calling into GetTextDocumentAsync and verifying its a `Document`, instead of implementing it again?

Reply:
Ah, sure, I can do that now. GetTextDocumentAsync didn't exist when I wrote the earlier bit.